addons/calendar_sms/models/calendar_alarm_manager.py

```diff
@@ -16,9 +16,10 @@ def _send_reminder(self):
         if not events_by_alarm:
             return
 
-        event_ids = list(set(event_id for event_ids in events_by_alarm.values() for event_id in event_ids))
-        events = self.env['calendar.event'].browse(event_ids)
         alarms = self.env['calendar.alarm'].browse(events_by_alarm.keys())
-        for event in events:
-            alarm = event.alarm_ids.filtered(lambda alarm: alarm.id in alarms.ids)
-            event._do_sms_reminder(alarm)
+        all_events_ids = list(set(event_id for event_ids in events_by_alarm.values() for event_id in event_ids))
+        for alarm in alarms:
+            event_ids = events_by_alarm.get(alarm.id, [])
+            events = self.env['calendar.event'].browse(event_ids).with_prefetch(all_events_ids)
+            for event in events:
+                event._do_sms_reminder(alarm)
```

Comment:
could directly do `events._do_sms_reminder(alarm)` so this ?

```suggestion
        all_events_ids = list(set(event_id for event_ids in events_by_alarm.values() for event_id in event_ids))
        for alarm, event_ids in events_by_alarms.items():
            events = self.env['calendar.event'].browse(event_ids).with_prefetch(all_events_ids)
            events._do_sms_reminder(alarm)
```

Reply:
I still need to retreive the alarm object so it would be this @jeh-odoo 

`for alarm_id, event_ids in events_by_alarm.items(): 
    alarm = self.env['calendar.alarm'].browse(alarm_id)
    events = self.env['calendar.event'].browse(event_ids).with_prefetch(all_events_ids)
    events._do_sms_reminder(alarm)`

Reply:
Ah yes indeed, I forgot about that, so yes something like this then (not tested)
```suggestion
        all_events_ids = list(set(event_id for event_ids in events_by_alarm.values() for event_id in event_ids))
        for alarm, event_ids in events_by_alarms.items():
            alarm = self.env['calendar.alarm'].browse(alarm).with_prefetch(events_by_alarm.keys())
            events = self.env['calendar.event'].browse(event_ids).with_prefetch(all_events_ids)
            events._do_sms_reminder(alarm)
```

Reply:
@jeh-odoo done and it works